### PR TITLE
feat: adds `store_inputs? true/false` option

### PR DESCRIPTION
**Pull request by chrishop**
closes: #20

- introduces `store_inputs?` to the DSL
- `store_inputs?` if not included defaults to false
- when switched on it ensure an attribute is created to capture the exact inputs of each new version, when the resource we are paper-trailing changes.

### CODE REVIEW

**Comment by chrishop** [comment]
@rgraff let me know if this is what you had in mind, look at the test. I think you should be able to tell by that.

**Review by zachdaniel — APPROVED** [review]
Will let @rgraff take a look a well.

File: lib/resource/resource.ex [hunk 1]
```diff
@@ -94,6 +94,13 @@ defmodule AshPaperTrail.Resource do
         doc: """
         Extensions that should be used by the version resource. For example: `extensions: [AshGraphql.Resource], notifier: [Ash.Notifiers.PubSub]`
         """
+      ],
+      store_inputs?: [
+        type: :boolean,
+        default: false,
+        doc: """
+        Whether or not to store the original inputs to the action as well as the changes within the version resource.
```
Comment (rgraff) [inline lib/resource/resource.ex:102]:
```suggestion
        Whether or not to store the original inputs to the action as well as the changes within the version resource.  Note: There's no filtering of the user input, so if your resource has password attributes or arguments, beware the will be stored in plain text.
```

File: lib/resource/resource.ex [hunk 2]
```diff
@@ -94,6 +94,13 @@ defmodule AshPaperTrail.Resource do
         doc: """
         Extensions that should be used by the version resource. For example: `extensions: [AshGraphql.Resource], notifier: [Ash.Notifiers.PubSub]`
         """
+      ],
+      store_inputs?: [
```
Comment (rgraff) [inline lib/resource/resource.ex:98]: Now that I think about the security ramifications of this, I'll probably want to expand it to:

```elixir
paper_trail do
  store_inputs? true, except: [:password]
end
```

But that can do in a future PR

**Review by rgraff — COMMENTED** [review]
I think this is good and would meet my use case. 🙇‍♂️  

I left some comments and questions.

File: lib/resource/changes/create_new_version.ex [hunk 3]
```diff
@@ -128,4 +130,12 @@ defmodule AshPaperTrail.Resource.Changes.CreateNewVersion do
     {:ok, dumped_value} = Ash.Type.dump_to_embedded(attribute.type, value, [])
     Map.put(changes, attribute.name, dumped_value)
   end
+
+  defp maybe_add_original_inputs(params, changeset) do
+    if AshPaperTrail.Resource.Info.store_inputs?(changeset.resource) do
+      Map.put(params, :inputs, Map.merge(changeset.arguments, changeset.attributes))
```
[the remark this answers is not in the document]
Comment (zachdaniel) [reply lib/resource/changes/create_new_version.ex:136]: Yeah, good point, this should be flipped.

However, we can't really use the raw params before casting unfortunately, as we have no way of knowing how to serialize them.

File: test/ash_paper_trail_test.exs [hunk 4]
```diff
@@ -260,4 +257,34 @@ defmodule AshPaperTrailTest do
...
+      assert AshPaperTrail.Resource.Info.store_inputs?(Posts.Post) == true
+
+      Posts.Post.create!(@valid_attrs, tenant: "acme")
+
+      post_version = Posts.Api.read!(Posts.Post.Version, tenant: "acme")
+
+      assert [
+               %{
+                 inputs: %{
+                   author: %AshPaperTrail.Test.Posts.Author{},
```
[the remark this answers is not in the document]
Comment (zachdaniel) [reply test/ash_paper_trail_test.exs:272]: Yep, you are correct.

File: lib/resource/resource.ex (hunk 1 above)
Comment (zachdaniel) [reply lib/resource/resource.ex:102]: We could actually scrub these. When storing the input, we could check the action, see if the value corresponds to a sensitive argument or sensitive attribute, and if so replace it with `**redacted**` in the changes? We would also need to traverse any unions/embeds given and scrub those too.

File: lib/resource/resource.ex (hunk 2 above)
Comment (zachdaniel) [reply lib/resource/resource.ex:98]: The scrubbing of sensitive inputs I mentioned above might solve for this specific case.

File: lib/resource/resource.ex (hunk 1 above)
Comment (rgraff) [reply lib/resource/resource.ex:102]: I like that approach. Makes dumping the input more difficult. @chrishop let me know if you want to do that or I can put it in another issue for later.

File: lib/resource/resource.ex (hunk 2 above)
Comment (rgraff) [reply lib/resource/resource.ex:98]: Yeah, I like the scrubbing of sensitive input better

File: lib/resource/resource.ex (hunk 1 above)
Comment (chrishop) [reply lib/resource/resource.ex:102]: From what @zachdaniel is saying it sounds like this needs to apply to `:changes` as well as `:inputs`, so I'm up for doing it later, but I think it should happen outside of this PR.

Comment (zachdaniel) [reply lib/resource/resource.ex:102]: If we don't scrub the values in this PR(which is fine) then we should mark the whole attribute as sensitive then :)

File: lib/resource/changes/create_new_version.ex (hunk 3 above)
[the remark this answers is not in the document]
Comment (chrishop) [reply lib/resource/changes/create_new_version.ex]: Flipped them around in this commit: [b16d054](https://github.com/ash-project/ash_paper_trail/pull/29/commits/b16d054c0c450f0df3f1ff53c6c17ca4309c5eaa)

This also recursively changes all of the Structs and casts them to a map using `Ash.Type.dump_to_embedded/3`

File: lib/resource/transformers/create_version_resource.ex [hunk 5]
```diff
@@ -162,6 +164,10 @@ defmodule AshPaperTrail.Resource.Transformers.CreateVersionResource do
             end
           end
 
+          if unquote(store_inputs?) do
+            attribute :inputs, :map
```
[the remark this answers is not in the document]
Comment (chrishop) [reply lib/resource/transformers/create_version_resource.ex]: I've made it sensitive in this commit: [5ad9e32](https://github.com/ash-project/ash_paper_trail/pull/29/commits/5ad9e324530164b77ecbfd59fb7503dd2804ee00)

File: lib/resource/resource.ex (hunk 1 above)
Comment (chrishop) [reply lib/resource/resource.ex:102]: made a new issue for it here: https://github.com/ash-project/ash_paper_trail/issues/30
will do a PR off the back of it.

File: lib/resource/resource.ex (hunk 2 above)
Comment (chrishop) [reply lib/resource/resource.ex:98]: made a new issue for it here: https://github.com/ash-project/ash_paper_trail/issues/30
will do a PR off the back of it.

File: lib/resource/transformers/create_version_resource.ex (hunk 5 above)
[the remark this answers is not in the document]
Comment (chrishop) [reply lib/resource/transformers/create_version_resource.ex]: issue is here:  https://github.com/ash-project/ash_paper_trail/issues/30

File: test/ash_paper_trail_test.exs (hunk 4 above)
[the remark this answers is not in the document]
Comment (chrishop) [reply test/ash_paper_trail_test.exs]: addressed here: [b16d054](https://github.com/ash-project/ash_paper_trail/pull/29/commits/b16d054c0c450f0df3f1ff53c6c17ca4309c5eaa)

**Comment by chrishop** [comment]
I think I addressed everything @rgraff.
I've made an issue to remove sensitive inputs from `:changes` and `:inputs` attributes and will get going on it.
#30

File: lib/resource/changes/create_new_version.ex [hunk 6]
```diff
@@ -128,4 +130,38 @@ defmodule AshPaperTrail.Resource.Changes.CreateNewVersion do
     {:ok, dumped_value} = Ash.Type.dump_to_embedded(attribute.type, value, [])
     Map.put(changes, attribute.name, dumped_value)
   end
+
+  defp maybe_add_original_inputs(params, changeset) do
+    if AshPaperTrail.Resource.Info.store_inputs?(changeset.resource) do
+      inputs =
+        changeset.attributes
+        |> Map.merge(changeset.arguments)
+        |> normalise_inputs()
```
Comment (zachdaniel) [inline lib/resource/changes/create_new_version.ex:139]: We shouldn't do this by looking for structs, we should get the corresponding argument or attribute and use `dump_to_embedded` on it regardless of if the value is a struct or not.

Comment (zachdaniel) [reply lib/resource/changes/create_new_version.ex:139]: That will take care of the recursive aspect as well.

**Review by zachdaniel — CHANGES_REQUESTED** [review]

File: lib/resource/changes/create_new_version.ex (hunk 6 above)
Comment (rgraff) [reply lib/resource/changes/create_new_version.ex:139]: Just this one open issue. Otherwise, looks good.

**Comment by zachdaniel** [comment]
Closing for now due to inactivity, happy to revisit at a later point.

**Pull request closed without merging**